File: client-python/examples/batch_operations.py
#!/usr/bin/env python3
"""
Batch operations example for soulseekR Python client
"""

import asyncio
from soulseekr import SoulseekrClient


async def main():
    """Batch operations example"""
    
    client = SoulseekrClient(
        base_url="http://localhost:8080",
        token="your-api-key-here",
        debug=True
    )
    
    try:
        # Create batch of operations
        batch = client.batch.builder()
        
        # Add multiple GET operations
        batch.get("/api/stats", op_id="get-stats")
        batch.get("/api/capabilities", op_id="get-caps")
        batch.get("/api/config", op_id="get-config")
        
        # Add search creation operations
        batch.post(
            "/api/searches",
            {"query": "bach concerto"},
            op_id="search-1"
        )
        batch.post(
            "/api/searches",
            {"query": "mozart sonata"},
            op_id="search-2"
        )
        
        print(f"Executing batch of {batch.size()} operations...")
        
        # Execute batch
        response = await batch.execute()
        
        print(f"\nBatch completed in {response.total_time_ms}ms")
        print(f"Successful: {len(response.get_successful())}")
        print(f"Failed: {len(response.get_failed())}")
        
        # Process results
        for result in response.results:
            print(f"\nOperation: {result.id}")
            print(f"  Status: {result.status}")
            if result.is_success():
                print(f"  Result: {str(result.body)[:100]}...")
            else:
                print(f"  Error: {result.body}")
        
        # Demonstrate batch monitoring
        if response.all_successful():
            print("\n✓ All operations successful!")
        else:
            print(f"\n✗ {len(response.get_failed())} operations failed")
            
    except Exception as e:
        print(f"Error: {e}")
    finally:
        await client.close()


if __name__ == "__main__":
    asyncio.run(main())

File: client-go/examples/batch_operations.go
//go:build examples

package main

import (
	"context"
	"fmt"
	"log"
	"time"

	"github.com/snapetech/slskr/client-go"
)

func main() {
	client := slskr.NewClient("http://localhost:8080", "your-api-key-here")
	client.HTTPClient.Timeout = 30 * time.Second

	ctx := context.Background()

	// Create batch builder
	batch := client.NewBatchBuilder()

	// Add multiple operations
	opID1 := "get-stats"
	opID2 := "get-caps"
	opID3 := "get-config"

	batch.Get("/api/stats", &opID1)
	batch.Get("/api/capabilities", &opID2)
	batch.Get("/api/config", &opID3)

	// Add search operations
	opID4 := "search-1"
	opID5 := "search-2"

	batch.Post("/api/searches", map[string]interface{}{
		"query": "bach concerto",
	}, &opID4)

	batch.Post("/api/searches", map[string]interface{}{
		"query": "mozart sonata",
	}, &opID5)

	fmt.Printf("Executing batch of %d operations...\n", batch.Size())

	// Execute batch
	response, err := batch.Execute(ctx)
	if err != nil {
		log.Fatalf("Error executing batch: %v", err)
	}

	fmt.Printf("\nBatch completed in %dms\n", response.TotalTimeMs)
	fmt.Printf("Successful: %d\n", len(response.GetSuccessful()))
	fmt.Printf("Failed: %d\n", len(response.GetFailed()))

	// Process results
	for _, result := range response.Results {
		fmt.Printf("\nOperation: %s\n", result.ID)
		fmt.Printf("  Status: %d\n", result.Status)
		if result.IsSuccess() {
			fmt.Printf("  Result: %v\n", result.Body)
		} else {
			fmt.Printf("  Error: %v\n", result.Body)
		}
	}

	// Check overall success
	if response.AllSuccessful() {
		fmt.Println("\n✓ All operations successful!")
	} else {
		fmt.Printf("\n✗ %d operations failed\n", len(response.GetFailed()))
	}
}
